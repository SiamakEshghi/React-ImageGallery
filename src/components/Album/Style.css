.secOne{
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 60px;
}
.searchForm {
    display: flex;
    align-items: center;
    background-color: white;
    margin-left: 10px;
    border-radius: 5px;
}
.searchForm * {
    border-width: 0px;
    height: 30px;
    background: transparent;
}
.input {
    width: 300px;
}
.select{
    margin-right: 10px;
    border-width: 2px;
    height: 30px;
    border-color: rgb(243, 219, 115);
}

.albumContainer {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
  }
  .albumContainer img {
    transition: all 2s;
  }
  .albumContainer img:hover {
    transform: scale(1.1,1.1);
  }
  .albumContainer div {
      width: 200px;
      height: 200px;
      margin: 5px;
  }
 .loadingContainer {
     display: flex;
     justify-content: center;
 }
 h2 {
     margin-left: 10px;
 }