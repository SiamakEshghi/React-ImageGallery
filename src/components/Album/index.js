import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Ionicon from 'react-ionicons';
import 'components/Album/Style.css';
import * as action from 'action';

const loadingGif = 'https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif';

class Album extends Component {
    state = {
        loaded: false,
        query: '',
        opt: 'all'
    }
    componentDidMount() {
        const { imageList } = this.props;
        if(imageList.length === 0){
            this.props.setImageList(this.state.query,() => {
                this.setState({ loaded: true });
            })
        }else {
            this.setState({ loaded: true });
        }
    }

   
    handleSearchRequest = event => {
        event.preventDefault();
        this.setState({ loaded: false}, () => {
            this.props.setImageList(this.state.query,() => {
                this.setState({ loaded: true, query: '' });
            });
        });
    }
    handleSelect = (event) => {
        const opt = event.target.value;
        this.setState({ opt })
    }
    handleChange = (event) => {
        const query = event.target.value;
        this.setState({ query })
    }
    openImage = (img) => {
        if(!img.myPhoto) {
            window.open(img.largeImageURL);
        } else {
            alert("Uploaded images are visible only in small size!")
        }
    }
    getFinalImageList = () => {
        const { opt } = this.state;
        const { imageList } = this.props;
        switch(opt) {
            case 'all':
                return imageList;
            case 'online':
                return imageList.filter((image) => !image.myPhoto);
            case 'upload':
                return imageList.filter((image) => image.myPhoto);
            default: return imageList;
        }
    }
    renderImages = () => {
        return (
            <div className="albumContainer">
                {this.getFinalImageList().map((image) => {
                    return (
                        <div key={image.id} >
                            <img 
                                onClick={() => this.openImage(image)} 
                                src={image.largeImageURL}
                            />
                        </div>
                    );
                })}
            </div>
        );   
    }
    renderLoading = () => {
        return <div className="loadingContainer"><img src={loadingGif}/></div>
    }
    render(){
        const { loaded, query } = this.state;
        return(
            <div>
                <div className="secOne">
                    <form 
                      className="searchForm" 
                      onSubmit={this.handleSearchRequest}
                    >
                        <input 
                            className="input" 
                            type="text" 
                            placeholder="Search.." 
                            name="search"
                            value={query} 
                            onChange={this.handleChange}
                        />
                        <button 
                         type="submit" 
                         className="searchBtn" 
                        >
                            <Ionicon icon="md-search"  color="gray" />
                        </button>
                    </form>
                    <select className="select" onChange={this.handleSelect} defaultValue="all">
                        <option value="all">-- Filter --</option>
                        <option value="online">Pixabay Images</option>
                        <option value="upload">Uploaded Images</option>
                    </select>
                </div>
               {loaded  ?  <div>
                                <h2>Images</h2>
                                {this.renderImages()}
                            </div>
                        : this.renderLoading()
                }
            </div>
        );
    }
}
const mapStateToProps = ({ img }) => {
    return { 
        imageList: img.imageList
    }
}
Album.propTypes = { imageList: PropTypes.array.isRequired };
export default connect(mapStateToProps, action)(Album);