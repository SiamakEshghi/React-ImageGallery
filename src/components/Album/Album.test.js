import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import Album from 'components/Album';

let wrapper, store, parentWrapper;
const middlewares = [reduxThunk];
const mockStore = configureStore(middlewares);
const initialState = {img: { imageList: [{id:1, largeImageURL: 'url1'}]} };


beforeEach(() => {
    store = mockStore(initialState);
    parentWrapper = shallow(<Album store={store} /> );
    wrapper = parentWrapper.dive();
});

afterEach(() => {
    wrapper.unmount();
});

describe('Album UI', () => {

    it('should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should show 1 form, 1 input, 1 button, 1 select', () => {
        expect(wrapper.find('form').length).toEqual(1);
        expect(wrapper.find('button').length).toEqual(1);
        expect(wrapper.find('select').length).toEqual(1);
        expect(wrapper.find('input').length).toEqual(1);
    });
    

    it('check Prop matches with initialState', () => {
        expect(parentWrapper.find('Album').prop('imageList')).toEqual(initialState.img.imageList)
    });

    it('should render an instance of the img component for each index of imageList', () => {
        const imageList = [
            {id: 2, largeImageURL: 'Url2'}, 
            {id: 3, largeImageURL: 'Url3'}
        ]
        wrapper.setProps({imageList});
        expect(wrapper.find('img').length).toEqual(2);
    });
});

describe('The Search Area', () => {
    beforeEach(() => {
        wrapper.find('input').simulate('change', {
            target: { value: 'new search'}
        });
        wrapper.update();
    });

    it('has a input that user can type in', () => {
        expect(wrapper.find('input').prop('value')).toEqual('new search');
    })
    
    it('after submitting the input is completly empty', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault: () => {}
        });
        wrapper.update();
        setTimeout(() => expect(wrapper.find('input').prop('value')).toEqual('')
        , 1000);
    });
});

describe('Select', () => {
    const imageList = [
        {id: 2, largeImageURL: 'Url2', myPhoto: true},
        {id: 2, largeImageURL: 'Url2'}, 
        {id: 3, largeImageURL: 'Url3'}
    ]
    beforeEach(() => {
        wrapper.setProps({imageList});
    });

    it('after selct Pixabay Images', () => {
        wrapper.find('select').simulate('change', {
            target: { value: 'online' }
        });
        expect(wrapper.find('img').length).toEqual(2);
    });
    
    it('after selct Uploaded Images', () => {
        wrapper.find('select').simulate('change', {
            target: { value: 'upload' }
        });
        expect(wrapper.find('img').length).toEqual(1);
    });
    it('after selct Filter Images', () => {
        wrapper.find('select').simulate('change', {
            target: { value: 'all' }
        });
        expect(wrapper.find('img').length).toEqual(3);
    });
});



