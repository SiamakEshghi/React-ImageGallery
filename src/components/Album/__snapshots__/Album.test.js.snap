// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album UI should render correctly 1`] = `
<div>
  <div
    className="secOne"
  >
    <form
      className="searchForm"
      onSubmit={[Function]}
    >
      <input
        className="input"
        name="search"
        onChange={[Function]}
        placeholder="Search.."
        type="text"
        value=""
      />
      <button
        className="searchBtn"
        type="submit"
      >
        <Ionicon
          beat={false}
          color="gray"
          fontSize="22px"
          icon="md-search"
          rotate={false}
          shake={false}
          style={Object {}}
        />
      </button>
    </form>
    <select
      className="select"
      defaultValue="all"
      onChange={[Function]}
    >
      <option
        value="all"
      >
        -- Filter --
      </option>
      <option
        value="online"
      >
        Pixabay Images
      </option>
      <option
        value="upload"
      >
        Uploaded Images
      </option>
    </select>
  </div>
  <div>
    <h2>
      Images
    </h2>
    <div
      className="albumContainer"
    >
      <div
        key="1"
      >
        <img
          onClick={[Function]}
          src="url1"
        />
      </div>
    </div>
  </div>
</div>
`;
